### PR TITLE
[Optimizer/TTNN] Migrating Optimizer to TTNN with minimal possible fu…

### PR DESCRIPTION
**Pull request by nobradovictt**
Migrating Optimizer to TTNN dialect. I tried to minimize functional changes but due to way how currently ToLayoutOp is handled in TTNN lowering and difference between layout attr in TTIR and TTNN dialect special handling was needed for memory management ops. Resharding is broken and will need to be rewritten, but based on inspecting old code it wouldn't have worked in all cases as well. These should be addressed in followup change in sync with changes to TTNN lowering and layout attr. 
All tests including Optimizer ones in passing state.
Pure Optimizer pass test has been updated to start from TTNN IR.

Closes https://github.com/tenstorrent/tt-mlir/issues/910

### CODE REVIEW

**Review by nsmithtt — APPROVED** [review]

**Comment by jnie-TT** [comment]
FYI @nobradovictt, this PR (#840) might be useful for the layout lowering and the current hacks we have in place.

File: lib/Dialect/TTNN/Transforms/Optimizer.cpp [hunk 1]
```diff
@@ -39,13 +37,17 @@ class TTIROptimizer : public impl::TTIROptimizerBase<TTIROptimizer> {
...
     ChipDescAttr chipDesc = systemDesc.getChipDescs()[0];
-    llvm::DenseMap<Operation *, std::vector<LayoutAttr>> legalLayouts;
+    llvm::DenseMap<Operation *, std::vector<tt::LayoutAttr>> legalLayouts;
 
     moduleOp->walk([&](Operation *op) {
       if (op->getNumResults() == 0) {
         return;
       }
 
+      if (!isa<RankedTensorType>(op->getResult(0).getType())) {
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/Transforms/Optimizer.cpp:47]: Why is this needed?

**Review by odjuricicTT — APPROVED** [review]

File: lib/Dialect/TTNN/Transforms/Optimizer.cpp (hunk 1 above)
Comment (nobradovictt) [reply lib/Dialect/TTNN/Transforms/Optimizer.cpp:47]: ttnn.get_device

**Comment by nobradovictt** [comment]
> FYI @nobradovictt, this PR (#840) might be useful for the layout lowering and the current hacks we have in place.

Thanks, Ill take a look, I need to make a followup change after this one regarding layout change handling in optimizer.

**Review by mtopalovicTT — APPROVED** [review]
